dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,261 @@
+        address participant;
+    }
+
+    // contractKey generates a key hash which commits to the contract data. The
+    // generated hash is used as a key in the swaps map.
+    function contractKey(address token, Vector calldata v) public pure returns (bytes32) {
+        return sha256(
+            bytes.concat(
+                v.secretHash,
+                bytes20(v.initiator),
+                bytes20(v.participant),
+                bytes32(v.value),
+                bytes8(v.refundTimestamp),
+                bytes20(token)
+            )
+        );
+    }
+
+    // Redemption is the information necessary to redeem a Vector. Since we
+    // don't store the Vector itself, it must be provided as part of the
+    // redemption.
+    struct Redemption {
+        Vector v;
+        bytes32 secret;
+    }
+
+    function secretValidates(bytes32 secret, bytes32 secretHash) public pure returns (bool) {
+        return sha256(bytes.concat(secret)) == secretHash;
+    }
+
+    // constructor is empty. This contract has no connection to the original
+    // sender after deployed. It can only be interacted with by users
+    // initiating, redeeming, and refunding swaps.
+    constructor() {}
+
+    // senderIsOrigin ensures that this contract cannot be used by other
+    // contracts, which reduces possible attack vectors.
+    modifier senderIsOrigin() {
+        require(tx.origin == msg.sender, "sender != origin");
+        _;
+    }
+
+    // retrieveStatus retrieves the current swap record for the contract.
+    function retrieveStatus(address token, Vector calldata v)
+        private view returns (bytes32, bytes32, uint256)
+    {
+        bytes32 k = contractKey(token, v);
+        bytes32 record = swaps[k];
+        return (k, record, uint256(record));
+    }
+
+    // status returns the current state of the swap.
+    function status(address token, Vector calldata v)
+        public view returns(Status memory)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveStatus(token, v);
+        Status memory r;
+        if (blockNum == 0) {
+            r.step = Step.Empty;
+        } else if (record == RefundRecord) {
+            r.step = Step.Refunded;
+        } else if (secretValidates(record, v.secretHash)) {
+            r.step = Step.Redeemed;
+            r.secret = record;
+        } else {
+            r.step = Step.Filled;
+            r.blockNumber =  blockNum;
+        }
+        return r;
+    }
+
+    // initiate initiates an array of Vectors.
+    function initiate(address token, Vector[] calldata contracts)
+        public
+        payable
+        senderIsOrigin()
+    {
+        uint initVal = 0;
+        for (uint i = 0; i < contracts.length; i++) {
+            Vector calldata v = contracts[i];
+
+            require(v.value > 0, "0 val");
+            require(v.refundTimestamp > 0, "0 refundTimestamp");
+            require(v.secretHash != RefundRecordHash, "illegal secret hash (refund record hash)");
```

Comment:
0 val and refund hash can be checked by client. They can't result in the initiator cheating.